context/app/static/js/hooks/useSortState.ts

```diff
@@ -0,0 +1,55 @@
+import { useCallback, useReducer } from 'react';
+
+type SortState = {
+  columnId?: string;
+  direction?: 'asc' | 'desc';
+};
+
+type SortAction =
+  | {
+      type: 'sort';
+      payload: string;
+    }
+  | {
+      type: 'reset';
+    };
+
+const initialSortState: SortState = {
+  columnId: undefined,
+  direction: undefined,
+} as SortState;
+
+function sortReducer(state: SortState, action: SortAction): SortState {
+  switch (action.type) {
+    case 'sort':
+      if (state.columnId === action.payload) {
+        return {
+          columnId: action.payload,
+          direction: state.direction === 'asc' ? 'desc' : 'asc',
+        };
+      }
+      return {
+        columnId: action.payload,
+        direction: 'desc',
+      } as SortState;
+    case 'reset':
+      return initialSortState;
+    default:
+      console.warn('Unrecognized action type', action);
+      return state;
+  }
+}
+
+export const useSortState = () => {
+  const [sortState, dispatch] = useReducer(sortReducer, initialSortState);
```

Comment:
Nice! What led you to use a reducer?

Reply:
I like to use reducers when there are interdependent state values, particularly when they share a control or are derived from a previous state. Using this approach lets us define the valid transitions explicitly to avoid accidentally causing an unintended state transition.

In this case, the new sort order is derived from a combination of the user-selected next state and the previous state:
- if the selected column is the same as the previous selected column, flip the existing sort order;
- if the selected column is not the same as the previous selected column, set the sort order to "desc"

Explicitly defining the actions that can be used to modify this sort state ensures we don't accidentally cause an unexpected invalid state with a `setState` call that violates those rules, and provides a more readable interface than a comparable `setState(previousState => ....)` call which would define this behavior inline.